stream_batch.go

```diff
@@ -0,0 +1,220 @@
+		config:         config,
+		generationTime: generationTime,
+		streams:        streams,
+		keyspaceName:   keyspaceName,
+		tableName:      tableName,
+
+		lastTimestamp: startFrom,
+
+		interruptCh: make(chan struct{}, 1),
+	}
+}
+
+func (sbr *streamBatchReader) run(ctx context.Context) error {
+	baseTableName := sbr.keyspaceName + "." + sbr.tableName
+
+	oldProgress := make(map[string]gocql.UUID, len(sbr.streams))
+	for _, stream := range sbr.streams {
+		progress, err := sbr.config.ProgressManager.GetProgress(sbr.generationTime, baseTableName, stream)
+		if err != nil {
+			return err
+		}
+		oldProgress[string(stream)] = progress.Time
+	}
+
+	minTime := oldProgress[string(sbr.streams[0])]
+	for _, progress := range oldProgress {
+		if compareTimeuuid(minTime, progress) > 0 {
+			minTime = progress
+		}
+	}
+	if (minTime != gocql.UUID{}) {
+		sbr.lastTimestamp = minTime
+	}
+
+	sbr.config.Logger.Printf("running batch starting from %v", sbr.lastTimestamp.Time())
+
+	input := CreateChangeConsumerInput{
+		TableName: baseTableName,
+		StreamIDs: sbr.streams,
+	}
+	consumer, err := sbr.config.ChangeConsumerFactory.CreateChangeConsumer(input)
+	if err != nil {
+		sbr.config.Logger.Printf("error while creating change consumer (will quit): %s", err)
+	} else {
+		defer consumer.End()
+	}
+
+	crq := newChangeRowQuerier(sbr.config.Session, sbr.streams, sbr.keyspaceName, sbr.tableName)
+
+	// sbr.config.Logger.Printf("starting stream processor loop for %v", sbr.streams)
+outer:
+	for {
+		timeWindowEnd := sbr.lastTimestamp.Time().Add(sbr.config.Advanced.QueryTimeWindowSize)
+		confidenceWindowEnd := time.Now().Add(-sbr.config.Advanced.ConfidenceWindowSize)
+
+		if timeWindowEnd.After(confidenceWindowEnd) {
+			timeWindowEnd = confidenceWindowEnd
+		}
+
+		pollEnd := gocql.MinTimeUUID(timeWindowEnd)
+
+		var (
+			err      error
+			rowCount int
+		)
+
+		readUpTo := pollEnd
+
+		if compareTimeuuid(sbr.lastTimestamp, pollEnd) < 0 {
+			// Set the time interval from which we need to return data
+			var iter *changeRowIterator
+			iter, err = crq.queryRange(sbr.lastTimestamp, pollEnd)
+			if err != nil {
+				sbr.config.Logger.Printf("error while sending a query (will retry): %s", err)
+			} else {
+				var change Change
+				for {
+					streamCols, c := iter.Next()
+					if c == nil {
+						break
+					}
+
+					if c.GetOperation() == PreImage {
+						change.Preimage = append(change.Preimage, c)
+					} else if c.GetOperation() == PostImage {
+						change.Postimage = append(change.Postimage, c)
+					} else {
+						change.Delta = append(change.Delta, c)
+					}
+
+					if c.cdcCols.endOfBatch {
+						// Prevent changes older than the save point from being replayed again
+						if compareTimeuuid(oldProgress[string(change.StreamID)], streamCols.time) < 0 {
+							change.StreamID = streamCols.streamID
+							change.Time = streamCols.time
+							if err := consumer.Consume(change); err != nil {
+								// TODO: Does that make sense?
```

Comment:
probably, but the API of `Consume` should state this precisely: if it returns an error, we will stop the entire thing.
Thus implementations of `Consume` should try at all cost to ensure that it succeeds. In particular if it can retry whatever it tries to do, it should

Also I wonder if it wouldn't make sense to implement some asynchronous consumption mechanism, so we can read the log further in parallel with consumption
e.g. instead of calling `Consume` and waiting here, we would put the change into a queue, and `Consume` would fetch from the queue? Of course the queue would be bounded in case `Consume` is slow

Not saying we have to do this now, but maybe as a future improvement

Reply:
also right now `Consume` for different consumers are called sequentially, while we could potentially call them concurrently...

Reply:
> probably, but the API of Consume should state this precisely: if it returns an error, we will stop the entire thing.

I'll definitely document this.

> Also I wonder if it wouldn't make sense to implement some asynchronous consumption mechanism, so we can read the log further in parallel with consumption

The problem is that the results from a poll are sorted first by stream_id, then by timestamp. We would have to read all rows from a poll first in order to parallelize processing within a stream batch.

> also right now Consume for different consumers are called sequentially, while we could potentially call them concurrently...

I think this is a good argument for making a consumer to be per-stream, not per-batch. If we decide to process streams within batch concurrently, then per-stream consumers could still process changes sequentially.

Reply:
Switch to a model with one consumer per stream.